In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('DataSets/ItemDataWithAiDescription.csv', index_col=0)

In [3]:
df

ProductName          Brand  Price  \
0           1017 ALYX 9SM Circle Logo Popover Hoodie  1017-alyx-9sm  325.0   
0                   1017 ALYX 9SM Side Zip Camo Pant  1017-alyx-9sm  443.0   
0            1017 ALYX 9SM x Blackmeans Denim Jacket  1017-alyx-9sm  779.0   
0   1017 ALYX 9SM Oversized Camo Nylon Bomber Jacket  1017-alyx-9sm  779.0   
0            1017 ALYX 9SM Nylon Logo Varsity Jacket  1017-alyx-9sm  489.0   
..                                               ...            ...    ...   
0                                YMC Embroidered Cap            ymc   62.0   
0                      YMC Ibiza '89 Pyramid T-Shirt            ymc   47.0   
0                      YMC Ibiza '89 Dancers T-Shirt            ymc   43.0   
0                       YMC Ibiza '89 Sunset T-Shirt            ymc   43.0   
0                                  Yogi Lennon Suede           yogi   95.0   

            Category Sizes                                    TextDescription  \
0   Hoodies & Sweats    []  Drawing influences from California skate cultu...   
0           Trousers    []  1017 ALYX 9SM innovative designs take cues fro...   
0    Coats & Jackets    []  Working together on a capsule that creates a r...   
0    Coats & Jackets    []  With an unmistakable experimental streetwear a...   
0    Coats & Jackets    []  Matthew Williams’s genre-blurring, future-orie...   
..               ...   ...                                                ...   
0        Accessories    []  You can never have too many caps, so add to yo...   
0           T-Shirts    []  YMC joins forces with Dave Swindells to celebr...   
0           T-Shirts    []  YMC continues to encourage us to create our ow...   
0           T-Shirts    []  Shot over a one week assignment for 20/20 maga...   
0           Footwear    []  One of Yogi’s  most recognised silhouettes, th...   

                               BulletPointDescription  \
0   ['100% Cotton', 'Drawstring Hood', 'Kangaroo P...   
0   ['100% Nylon', 'Elasticated Waist', 'Zip Closu...   
0   ['100% Cotton', 'Classic Collar', 'Button Clos...   
0   ['100% Nylon', 'Stand Up Collar', 'Snap Button...   
0   ['100% Polyester', 'Shawl Collar', 'Snap Butto...   
..                                                ...   
0   ['74% Cotton, 26% Linen', '8-Wale Corduroy', '...   
0   ['100% Organic Slub Cotton', 'Oversized Fit', ...   
0   ['100% Organic Slub Cotton', 'Oversized Fit', ...   
0   ['100% Organic Slub Cotton', 'Oversized Fit', ...   
0   ['Suede Uppers', 'Natural Rubber Outsoles', 'W...   

                                            ImageUrls  \
0   ['https://media.endclothing.com/media/f_auto,q...   
0   ['https://media.endclothing.com/media/f_auto,q...   
0   ['https://media.endclothing.com/media/f_auto,q...   
0   ['https://media.endclothing.com/media/f_auto,q...   
0   ['https://media.endclothing.com/media/f_auto,q...   
..                                                ...   
0   ['https://media.endclothing.com/media/f_auto,q...   
0   ['https://media.endclothing.com/media/f_auto,q...   
0   ['https://media.endclothing.com/media/f_auto,q...   
0   ['https://media.endclothing.com/media/f_auto,q...   
0   ['https://media.endclothing.com/media/f_auto,q...   

                                        AIDescription  
0   The item is a hoodie from the category of Hood...  
0   The item is categorized as trousers and is spe...  
0   The item is a denim jacket categorized under C...  
0   The item is an **Oversized Camo Nylon Bomber J...  
0   The item is a coat categorized as a varsity ja...  
..                                                ...  
0   The YMC Embroidered Cap falls under the catego...  
0   The YMC Ibiza '89 Pyramid T-Shirt falls under ...  
0   The YMC Ibiza '89 Dancers T-Shirt is categoriz...  
0   The YMC Ibiza '89 Sunset T-Shirt is categorize...  
0   The Yogi Lennon Suede is a pair of footwear ca...  

[14912 rows x 9 columns]

In [12]:
product_descriptors = {
    "product_name": "The name of the item",
    "category": "The category, like t-shirt, trousers, jeans, skirt, long sleeve t-shirt, jacket, boots, dress shoes etc.",
    "brand": "Brand Name",
    "fashion_season": "The season the item/collection came out. Mostly applicable to high end fashion. E.g., Spring/Summer 2024 is SS24 or Autumn/Winter 2024 is AW24",
    "primary_colour": "The primary colour",
    "secondary_colour": "The secondary colour",
    "pattern": "The pattern on the clothing like floral, striped, plain etc.",
    "fabric": "The material of the fabric; e.g., cotton, polyester, wool",
    "price": "This is the price of the item",
    "price_category": "budget, mid-range, luxury, or haut couture",
    "size": "sizing of the product",
    "fit": "the fit of the item, e.g., tight, slim fit, regular, oversized, boxy",
    "gender_audience": "Is this aimed at men, women, or unisex",
    "age_audience": "Is this aimed at toddlers, children, teens, adults, or old aged.",
    "season": "What season is this aimed to be worn in: i.e., summer, winter, autumn, spring",
    "occasion": "Occasions or settings (e.g., casual, formal, athletic).",
    "sleeve_length": "For tops/jackets (e.g., short, long, sleeveless).",
    "neckline": "The style of the neckline (e.g., V-neck, crew neck).",
    "country_of_manufacture": "Where the item was made.",
    "closure_type": "Details such as buttons, zippers, hooks, or laces.",
    "pocket_details": "Specifics on the number, placement, or style of pockets",
    "embellishments": "Information on additional decorative elements like embroidery, sequins, or patches.",
    "second_hand": "True if the item is pre-owned",
    "description": "A textual description that highlights special features or details.",
    "image_urls": "Image urls",
    "sustainability_info": "Information on eco-friendliness, ethical production, or sustainability certifications.",
    "additional_features": "Any other unique details like waterproofing, stretchability, or tech integration. that haven't been described by the attributes above."
}

Some ideas for building the most accurate retriever

- Using the feature dictionary as above. An LLM can use description analysis and image analysis to populate a dictionary for the item with words/phrases as each value using the above reference dictionary as an example.

- I could force the LLM to choose between specific words for each key-value pair and then use a simple search for specific options when finding a similar item.

- Or I could use short embeddings for description words and use a similarity check so that descriptors that are close to eachother come up in the search. E.g., very expensive, expensive, and cheap could be one of 5 categories but I would prefer that an item in the very expensive category is more "similar" than the "cheap" one. This is especially applicable to descriptors of style where the lines are slightly more blurred and natural language would be better than specific options.

- I may want to use an AI agent to look at the customer description to assign weights (or even hard stops) on features. For exmaple if a user asks for shorts specifically in the query then the agent sets a hard filter on the catgegory to match shorts. Or also if a person says something under £20 then we only want to show items under £20.

- When I retrieve some items I might want to use an AI agent to have a look at the recommendations and rank them in usefullness compared to the user query as an additional check at the end.

- I could include a 'addtional comments' bit of the description where I ask the AI to give a brief description of the style of the item and any other information that might be useful and use a vector store on that specific column in conjunction with the other search functions.

- An important aspect will be how to handle features of the item dictionary which are null. I won't be able to fill every feature with the user query. Maybe only comparing on a certain subset of features which can actually be determined from the query.

In [15]:
item = df.iloc[0]

image_urls_as_list = item['ImageUrls'].replace("[", "").replace("]", "").replace("'", "").split(", ")
item_dict = {
    "Product Name": item["ProductName"],
    "Brand": item["Brand"],
    "Price": item["Price"],
    "Category": item["Category"],
    "TextDescription": item["TextDescription"],
    "BulletPointDescription": item["BulletPointDescription"]
}

In [14]:
image_urls_as_list

['https://media.endclothing.com/media/f_auto,q_auto:eco,w_200/prodmedia/media/catalog/product/1/4/14-04-2023-JA_AAUSW0177FA01-BLK0001_1_1.jpg',
 'https://media.endclothing.com/media/f_auto,q_auto:eco,w_200/prodmedia/media/catalog/product/1/4/14-04-2023-JA_AAUSW0177FA01-BLK0001_7_1.jpg',
 'https://media.endclothing.com/media/f_auto,q_auto:eco,w_200/prodmedia/media/catalog/product/1/4/14-04-2023-JA_AAUSW0177FA01-BLK0001_8_1.jpg',
 'https://media.endclothing.com/media/f_auto,q_auto:eco,w_200/prodmedia/media/catalog/product/1/4/14-04-2023-JA_AAUSW0177FA01-BLK0001_m9_1.jpg',
 'https://media.endclothing.com/media/f_auto,q_auto:eco,w_200/prodmedia/media/catalog/product/1/4/14-04-2023-JA_AAUSW0177FA01-BLK0001_10_1.jpg']

In [ ]:
# Using OpenAIs gpt40-mini to extract features from images and the table data for an item
from openai import OpenAI

client = OpenAI(api_key="")

response = client.chat.completions.create(
    model="gpt-4o-mini",
    messages=[
        {
            "role": "system",
            "content": f"""You are tasked with extracting features for an item of clothing. 
            You will be provided with images of the item and some textual data from a database. 
            This is a dictionary of the features (keys) with a brief description of the feature (values): 
            \n\n{str(product_descriptors)}\n\n You should respond with only valid JSON where the keys are the keys from the dictionary just shown. If there is no valid value for a certain feature just put an empty string ('')."""
        },
        {
            "role": "user",
            "content": [
                {
                    "type": "text",
                    "text": f"This is the data on the item from the database: {str(item_dict)}",
                },
                {
                    "type": "image_url",
                    "image_url": {
                        "url": image_urls_as_list[0],
                    },
                },
                {
                    "type": "image_url",
                    "image_url": {
                        "url": image_urls_as_list[1],
                    },
                },
                {
                    "type": "image_url",
                    "image_url": {
                        "url": image_urls_as_list[2],
                    },
                },
                {
                    "type": "image_url",
                    "image_url": {
                        "url": image_urls_as_list[3],
                    },
                },
                {
                    "type": "image_url",
                    "image_url": {
                        "url": image_urls_as_list[4],
                    },
                }
            ],
        }
    ],
    max_tokens=500,
)

print(response.choices[0])

Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='```json\n{\n    "product_name": "1017 ALYX 9SM Circle Logo Popover Hoodie",\n    "category": "Hoodies & Sweats",\n    "brand": "1017-alyx-9sm",\n    "fashion_season": "",\n    "primary_colour": "Black",\n    "secondary_colour": "White",\n    "pattern": "",\n    "fabric": "100% Cotton",\n    "price": "325.0",\n    "price_category": "luxury",\n    "size": "",\n    "fit": "regular",\n    "gender_audience": "unisex",\n    "age_audience": "adults",\n    "season": "autumn",\n    "occasion": "casual",\n    "sleeve_length": "long",\n    "neckline": "hooded",\n    "country_of_manufacture": "Italy",\n    "closure_type": "drawstring",\n    "pocket_details": "kangaroo pocket",\n    "embellishments": "circle logo",\n    "second_hand": "",\n    "description": "Drawing influences from California skate culture and New York grit, the creative mind of Matthew Williams continues to experiment with each of 1017 ALY

In [27]:
string_output_cheap = response.choices[0].message.content

In [ ]:
string_output = response.choices[0].message.content

In [29]:
import json

def transform_string_json_to_dict(json_str: str):

    lines = string_output.splitlines()
    if lines[0].startswith("```") and lines[-1].startswith("```"):
        json_str = "\n".join(lines[1:-1])
    else:
        json_str = string_output

    # Convert the JSON string to a Python dictionary
    data = json.loads(json_str)

    return data

In [30]:
dict_output = transform_string_json_to_dict(string_output)
dict_output_cheap = transform_string_json_to_dict(string_output_cheap)

In [31]:
# Checking the output
output_dict_keys = set(dict_output_cheap.keys())
reference_dict_keys = set(product_descriptors.keys())

print(output_dict_keys.difference(reference_dict_keys))
print(reference_dict_keys.difference(output_dict_keys))

set()
set()


In [41]:
def compare_dicts(dict1, dict2):
    differences = {}
    # Create a set of all keys present in either dictionary
    all_keys = set(dict1.keys()) | set(dict2.keys())
    
    for key in all_keys:
        val1 = dict1.get(key)
        val2 = dict2.get(key)
        # If the values differ (or key is missing in one dictionary), record the difference
        if val1 != val2:
            differences[key] = [val1, val2]
    return differences

In [42]:
compare_dicts(dict_output, dict_output_cheap)

{}

Using 1 image is roughly 9.5k tokens or 0.15*(9500/1000000) = $0.0014
Using all 5 images is roughly 43k tokens or 0.15*(9500/1000000) = $0.00645

With 14912 items in my table it would cost me:
Using 1 image -> 14912*0.0014 = $20.88
Using all images (assuming a 5 image average) -> 14912*0.00645 = $96.18

For this 1 example using 1 image gave exactly the same output dictionary as using all the images so doesn't make sense to use all images (plus it  took 1s vs 5s so would increase the time taken by a lot)